### ISSUE TITLE
SuppressEntryChunks failing when cache-busting or hot-reload is used

### ISSUE THREAD

**Issue by actra-gschuster** [angular/angular-cli#6586]
### Bug Report or Feature Request (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
```

### Versions.
Everything up-to-date / latest versions as of June 2017.

### Repro steps.
Enable cache-busting (option "hash") for HtmlWebpackPlugin and check the outputted HTML - wrong <script>-tag is still there as the reg-ex used only matches when the src ends in ".css" but not for ".css?<cache-busting-hash>".

Additionally, in dev-mode with hot-reload, the check doesn't find the "all css"-chunks as all chunks get the webpack-dev-server client and hpt-reload unshifted.
Example for my polyfills-chunk:
[path-to]\node_modules\webpack-dev-server\client\index.js?http://localhost:4000,webpack/hot/dev-server,[path-to]\src\polyfills.ts

### Desired functionality.
Cache-busted entries should also be removed and webpack-dev-server unshifted entries ignored.

### Quick fix:
```
Line 14: if (entryPoints[entryPoint].every((el) => el.match(/(\.(css|scss|sass|less|styl)|node_modules[\\\/]webpack-dev-server[\\\/]client[\\\/]index\.js(\?https?\:\/\/.+)?|webpack[\\\/]hot[\\\/]dev-server)$/))) {

Line 25: if (file.match(/\.js(\.map)?(\?[a-z0-9]+)?$/)) {

Line 40: const filterFn = (tag) => !(tag.tagName === 'script' && tag.attributes.src.match(/\.css(\?[a-z0-9]+)?$/));
```

**Comment by actra-gschuster** [comment]
#6587

**Comment by clydin** [comment]
The CLI natively supports cache busting.  Is there a reason you're modifying the settings of `HtmlWebpackPlugin`?  Also of note is that the query string approach is not a 100% solution and may not work as intended in all environments.

**Comment by filipesilva** [comment]
I cannot reproduce what you're describing. This is what I did:
- `ng serve --output-hashing=all -o`
- in browser, inspect css/js bundle hashes
- make a change in css/ts files
- browser reloads, reflects changes
- bundle hashes also change

**Comment by actra-gschuster** [comment]
I'm not modifying any settings, I just enabled "hash = true" in the options passed to HtmlWebpackPlugin and with that setting my pure-css-entry is not suppressed as expected.

```
const { SuppressExtractedTextChunksWebpackPlugin } = require('@angular/cli/plugins/webpack');
// ...
let configuration = {
    entry:         {
        'main':      [
            helpers.root('src', 'main.ts')
        ],
        'polyfills': [
            helpers.root('src', 'polyfills.ts')
        ],
        'styles': [
            helpers.root('node_modules', 'primeng', 'resources', 'themes', 'omega', 'theme.scss'),
            helpers.root('node_modules', 'primeng', 'resources', 'primeng.css'),
            helpers.root('node_modules', 'fullcalendar', 'dist', 'fullcalendar.css'),
            helpers.root('node_modules', 'fullcalendar-scheduler', 'dist', 'scheduler.css')
        ]
    },
    // ...
    plugins:       [
            new HtmlWebpackPlugin({
                cache:          false,
                chunks:         'all',
                chunksSortMode: function sort(left, right) {
                    let leftIndex  = entryPoints.indexOf(left.names[0]);
                    let rightIndex = entryPoints.indexOf(right.names[0]);

                    return leftIndex > rightIndex ? 1 : (leftIndex < rightIndex ? -1 : 0);
                },
                compile:        true,
                environment:    constants.ENV,
                favicon:        false,
                filename:       'index.html',
                hash:           true,
                inject:         true,
                showErrors:     !!(constants.DEVELOPMENT || constants.TESTING),
                template:       helpers.root('src/index.html'),
                title:          constants.METADATA.application.title,
                xhtml:          true
            }),
            new ExtractTextPlugin({
                'filename': 'assets/css/[name].[contenthash:20].bundle.css',
                'disable':   false
            }),
            new SuppressExtractedTextChunksWebpackPlugin(),
    ],
    // ...
```
With the above configuration I would expect that there is NO <script>-tag referencing "assets/css/styles.[hash].bundle.css" - but THERE IS a script tag.
SuppressExtractedTextChunksWebpackPlugin doesn't catch it because of the "hash = true", if set to "hash = false" everything works as expected.

**Comment by clydin** [comment]
You don't appear to be using the CLI.  Is this an ejected config?

Also, it appears you are already adding hashes to your assets.  The HtmlWebpackPlugin  hash option provides no real benefit in that scenario and may in fact be counter-productive; as it can overzealously bust a user's cache.

Overall, I suppose allowing for a query and/or fragment in general could allow for a more complete solution.

**Comment by actra-gschuster** [comment]
Yes, it is an ejected config and I already use hashes in the filenames but it was a project requirement to add question-mark hashes, too, this is why I stumbled upon the problem.
As SuppressExtractedTextChunksWebpackPlugin may be used without the CLI I'd agree that it should allow query and/or fragment cache-busting, that's why I came up with the given solution.

**Comment by filipesilva** [comment]
Oh I see... this is using SuppressExtractedTextChunksWebpackPlugin separately. I agree it should be fixed, but the big regex in https://github.com/angular/angular-cli/pull/6587 looks like a nightmare to maintain. I'll  comment there.

**Comment by tiaguinho** [comment]
@filipesilva There is anything I can do to help fixing this?

**Comment by clydin** [comment]
Eject doesn’t exist anymore and the CLI no longer uses html-webpack-plugin; so this issue has most likely been overcome by those events.

However,  it would be worth investigating whether the suppress plugin is still actually needed with recent versions of webpack and the mini-css-extract-plugin.  Best case is the plugin is no longer needed.  Worst case the file can be cleaned up.
